File: previous/167.two-sum-ii-input-array-is-sorted.js
/*
 * @lc app=leetcode id=167 lang=javascript
 *
 * [167] Two Sum II - Input Array Is Sorted
 * Time Complexity: O(n), worst case, 兩數之和在陣列最中間
 * Space Complexity: O(1)
 */

// @lc code=start
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (numbers, target) {
	let p1 = 0,
		p2 = numbers.length - 1;

	while (true) {
		const twoSum = numbers[p1] + numbers[p2];
		if (twoSum > target) {
			p2--;
		} else if (twoSum < target) {
			p1++;
		} else {
			return [p1 + 1, p2 + 1];
		}
	}
};
// @lc code=end

File: 7-day-sprint/day3/167.two-sum-ii-input-array-is-sorted.py
#
# @lc app=leetcode id=167 lang=python3
#
# [167] Two Sum II - Input Array Is Sorted
#

# @lc code=start
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left, right = 0, len(numbers) - 1
        
        while(left <= right):
            if numbers[left] + numbers[right] == target:
                return [left + 1, right + 1]
            elif numbers[left] + numbers[right] > target:
                right -= 1
            else:
                left += 1

        return None
# @lc code=end

# Test Case
# [2,7,11,15]
# 9
# [2,3,4]
# 6
# [-1,0]
# -1

# Concept: Two pointer & Binary Search, 利用排序好的特性有策略的選擇 pointer 的移動
# Time: O(n), Space: O(1)